lib/api/api_test.go

```diff
@@ -566,60 +566,354 @@ func TestHTTPLogin(t *testing.T) {
 	if resp.StatusCode != http.StatusUnauthorized {
 		t.Errorf("Unexpected non-401 return code %d for unauthed request", resp.StatusCode)
 	}
+	if assertHasSessionCookie(resp.Cookies()) {
+		t.Errorf("Unexpected session cookie for unauthed request")
+	}
 
 	// Verify that incorrect password is rejected
-
-	req.SetBasicAuth("üser", "rksmrgs")
-	resp, err = http.DefaultClient.Do(req)
-	if err != nil {
-		t.Fatal(err)
-	}
+	resp = performRequest("üser", "rksmrgs") // string literals in Go source code are in UTF-8
 	if resp.StatusCode != http.StatusUnauthorized {
 		t.Errorf("Unexpected non-401 return code %d for incorrect password", resp.StatusCode)
 	}
+	if assertHasSessionCookie(resp.Cookies()) {
+		t.Errorf("Unexpected session cookie for incorrect password")
+	}
 
 	// Verify that incorrect username is rejected
+	resp = performRequest("user", "räksmörgås") // string literals in Go source code are in UTF-8
+	if resp.StatusCode != http.StatusUnauthorized {
+		t.Errorf("Unexpected non-401 return code %d for incorrect username", resp.StatusCode)
+	}
+	if assertHasSessionCookie(resp.Cookies()) {
+		t.Errorf("Unexpected session cookie for incorrect username")
+	}
 
-	req.SetBasicAuth("user", "räksmörgås") // string literals in Go source code are in UTF-8
-	resp, err = http.DefaultClient.Do(req)
+	// Verify that UTF-8 auth works
+	resp = performRequest("üser", "räksmörgås") // string literals in Go source code are in UTF-8
+	if resp.StatusCode != http.StatusOK {
+		t.Errorf("Unexpected non-200 return code %d for authed request (UTF-8)", resp.StatusCode)
+	}
+	if !assertHasSessionCookie(resp.Cookies()) {
+		t.Errorf("Expected session cookie for authed request (UTF-8)")
+	}
+
+	// Verify that ISO-8859-1 auth works
+	resp = performRequest("\xfcser", "r\xe4ksm\xf6rg\xe5s") // escaped ISO-8859-1
+	if resp.StatusCode != http.StatusOK {
+		t.Errorf("Unexpected non-200 return code %d for authed request (ISO-8859-1)", resp.StatusCode)
+	}
+	if !assertHasSessionCookie(resp.Cookies()) {
+		t.Errorf("Expected session cookie for authed request (ISO-8859-1)")
+	}
+}
+
+func TestHTTPLoginAtNotFoundPath(t *testing.T) {
+	t.Parallel()
+
+	cfg := newMockedConfig()
+	cfg.GUIReturns(config.GUIConfiguration{
+		User:     "üser",
+		Password: "$2a$10$IdIZTxTg/dCNuNEGlmLynOjqg4B1FvDKuIV5e0BB3pnWVHNb8.GSq", // bcrypt of "räksmörgås" in UTF-8
+		SendBasicAuthPrompt: true,
+	})
+	baseURL, cancel, err := startHTTP(cfg)
 	if err != nil {
 		t.Fatal(err)
 	}
+	defer cancel()
+	url := baseURL + "/any-path/that/does/nooooooot/match-any/noauth-pattern"
+
+	performRequest := func (username string, password string) *http.Response {
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if username != "" || password != "" {
+			req.SetBasicAuth(username, password)
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		return resp
+	}
+
+	assertHasSessionCookie := func (cookies []*http.Cookie) bool {
```

Comment:
Another potential for reuse. Not commenting any further.

Reply:
Yep, done!